paper-api/src/main/java/io/papermc/paper/registry/data/variant/SpawnCondition.java

```diff
@@ -0,0 +1,127 @@
+package io.papermc.paper.registry.data.variant;
+
+import com.google.common.collect.Range;
+import io.papermc.paper.registry.RegistryKey;
+import io.papermc.paper.registry.TypedKey;
+import io.papermc.paper.registry.set.RegistryKeySet;
+import io.papermc.paper.registry.set.RegistrySet;
+import org.bukkit.block.Biome;
+import org.bukkit.generator.structure.Structure;
+import org.jetbrains.annotations.Contract;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * A condition checked when an entity is trying to spawn.
+ */
+public sealed interface SpawnCondition permits SpawnCondition.BiomeCheck, SpawnCondition.MoonBrightnessCheck, SpawnCondition.StructureCheck {
+
+    /**
+     * A condition that checks if the position matches a biome.
+     *
+     * @param requiredBiomes The biomes that the position must match
+     * @return The biome check condition
+     */
+    @SafeVarargs
+    @Contract(pure = true, value = "_ -> new")
+    static SpawnCondition biomeCheck(final TypedKey<Biome>... requiredBiomes) {
+        return biomeCheck(RegistrySet.keySet(RegistryKey.BIOME, requiredBiomes));
+    }
+
+    /**
+     * A condition that checks if the position matches a biome.
+     *
+     * @param requiredBiomes The biomes that the position must match
+     * @return The biome check condition
+     */
+    @Contract(pure = true, value = "_ -> new")
+    static SpawnCondition biomeCheck(final RegistryKeySet<Biome> requiredBiomes) {
+        return new BiomeCheckImpl(requiredBiomes);
+    }
+
+    /**
+     * A condition that checks if the moon brightness is within a certain range.
+     *
+     * @param min The minimum moon brightness, or null if there is no minimum
+     * @param max The maximum moon brightness, or null if there is no maximum
+     * @return The moon brightness check condition
+     */
+    @Contract(pure = true, value = "_, _ -> new")
+    static SpawnCondition moonBrightnessCheck(final @Nullable Double min, final @Nullable Double max) {
```

Comment:
Since Range is already used for brigadier, might be okay to use guava again?

Reply:
no, I don't think so for "construction". The guava range has more possible values than the vanilla MinMaxBounds.Doubles. For example, in guava ranges, you can have "open" ranges. That isn't possible to represent in MinMaxBounds.Doubles, they are all closed ranges.

Reply:
You can check that in a precondition i think.

Reply:
yeah, alright.